lib/nvme/nvme_pcie.c

```diff
@@ -1817,36 +1819,37 @@ static void
 nvme_pcie_qpair_check_timeout(struct spdk_nvme_qpair *qpair)
 {
 	uint64_t t02;
-	struct nvme_tracker *tr;
+	struct nvme_tracker *tr, *tmp;
 	struct nvme_pcie_qpair *pqpair = nvme_pcie_qpair(qpair);
 	struct spdk_nvme_ctrlr *ctrlr = qpair->ctrlr;
 
-	if (TAILQ_EMPTY(&pqpair->outstanding_tr)) {
-		return;
+	/* We don't want to expose the admin queue to the user,
+	 * so when we're timing out admin commands set the
+	 * qpair to NULL.
+	 */
+	if (qpair == ctrlr->adminq) {
+		qpair = NULL;
 	}
 
-	/*
-	 * qpair could be either for normal i/o or for admin command. If qpair is admin
-	 * and request is SPDK_NVME_OPC_ASYNC_EVENT_REQUEST, skip to next previous.
-	 */
-	tr = TAILQ_LAST(&pqpair->outstanding_tr, nvme_outstanding_tr_head);
-	while (tr->req->cmd.opc == SPDK_NVME_OPC_ASYNC_EVENT_REQUEST) {
-		/* qpair is for admin request */
-		tr = TAILQ_PREV(tr, nvme_outstanding_tr_head, tq_list);
-		if (!tr) {
-			/*
-			 * All request were AER
+	t02 = spdk_get_ticks();
+	TAILQ_FOREACH_SAFE(tr, &pqpair->outstanding_tr, tq_list, tmp) {
```

Comment:
Ben, i think what you want is the oldest item but TAILQ_FOREACH_SAFE(..) assigns the youngest item to tr instead.

To get the oldest item each time in a loop, you can replace the TAILQ_FOREACH_SAFE(..) on line 1835 with this
for(tr = TAILQ_LAST(&pqpair->outstanding_tr, nvme_outstanding_tr_head); tr; tr = TAILQ_PREV(tr, nvme_outstanding_tr_head, tq_list)) {